src/test/mod.rs

```diff
@@ -417,6 +417,68 @@ fn test_workspace_symbol() {
                                                                      .expect_contains(r#""containerName":"foo""#)]);
 }
 
+#[test]
+fn test_workspace_symbol_duplicates() {
+    let mut env = Environment::new("workspace_symbol_duplicates");
+
+    let root_path = env.cache.abs_path(Path::new("."));
+
+    let messages = vec![
+        initialize(0, root_path.as_os_str().to_str().map(|x| x.to_owned())).to_string(),
+        request::<requests::WorkspaceSymbol>(
+            42,
+            WorkspaceSymbolParams {
+                query: "Frobnicator".to_owned(),
+            },
+        ).to_string(),
+    ];
+
+    env.with_config(|c| c.cfg_test = true);
+    let (mut server, results) = env.mock_server(messages);
+    // Initialize and build.
+    assert_eq!(
+        ls_server::LsService::handle_message(&mut server),
+        ls_server::ServerStateChange::Continue
+    );
+    expect_messages(
+        results.clone(),
+        &[
+            ExpectedMessage::new(Some(0)).expect_contains("capabilities"),
+            ExpectedMessage::new(None).expect_contains("progress").expect_contains(r#"title":"Building""#),
+            ExpectedMessage::new(None).expect_contains("progress").expect_contains("workspace_symbol"),
+            ExpectedMessage::new(None).expect_contains("progress").expect_contains(r#""done":true"#),
+            ExpectedMessage::new(None).expect_contains("progress").expect_contains(r#"title":"Indexing""#),
+            ExpectedMessage::new(None).expect_contains("progress").expect_contains(r#""done":true"#),
+        ],
+    );
+
+    assert_eq!(
+        ls_server::LsService::handle_message(&mut server),
+        ls_server::ServerStateChange::Continue
+    );
+
+    wait_for_n_results!(1, results);
+    let result: serde_json::Value =
+        serde_json::from_str(&results.lock().unwrap().remove(0)).unwrap();
+    let mut result = result.get("result").unwrap().clone();
+    *result.pointer_mut("/0/location/uri").unwrap() = "shared.rs".into();
```

Comment:
What do you think about bringing something like https://github.com/rust-lang/cargo/blob/5699afe508d62924f6b38b19dc98296ad33d1659/tests/testsuite/cargotest/support/mod.rs#L950-L1003 to check JSON messages instead of using `ExpectedMessage`? 

That is, specify expected result as JSON, with support for wildcards for places which you don't want to match.

Reply:
I think the `expect_contains` stuff should be totally replaced. 

Personally I prefer using the `json` crate for this ie `assert_eq!(deglob_loc["start"]["line"], 12);` as it's precise and simple. But any/other improvements would be appreciated.